File: src/conventions/pipeline/evaluation/alert.ts
import type { BidAlert } from "../../core/strategy-types";
import { Disclosure, type BidMeaningClause } from "./meaning";
import type { TeachingLabel } from "../../core/authored-text";

// ─── Disclosure-based conventionality derivation ─────────────────────
//
// Alertability is derived from the `disclosure` field on BidMeaning.
// Constraints carry `isPublic` on the FactConstraint itself; no separate
// publicConstraints pipeline is needed.

/** Minimal shape for alert resolution — only needs disclosure and teachingLabel.
 *  Alertability is derived from the `disclosure` field. */
export interface AlertResolvable {
  readonly disclosure: Disclosure;
  readonly clauses: readonly BidMeaningClause[];
  readonly teachingLabel: TeachingLabel;
}

/** Derive whether a bid is alertable from its disclosure classification.
 *  A bid is alertable (requires some form of opponent notification) unless
 *  it has natural meaning (disclosure === "natural"). */
export function isAlertable(
  disclosure: Disclosure,
): boolean {
  return disclosure !== Disclosure.Natural;
}

/** Resolve whether a surface is alertable.
 *  Returns BidAlert when the bid is conventional, null when natural.
 *  Alertability is derived from the `disclosure` field.
 *  annotationType distinguishes ACBL alerts, announcements, and educational labels. */
export function resolveAlert(surface: AlertResolvable): BidAlert | null {
  if (!isAlertable(surface.disclosure)) return null;

  const annotationType = surface.disclosure === Disclosure.Announcement
    ? "announce" as const
    : surface.disclosure === Disclosure.Standard
      ? "educational" as const
      : "alert" as const;

  return {
    teachingLabel: surface.teachingLabel.name,
    annotationType,
  };
}

File: crates/bridge-conventions/src/pipeline/evaluation/alert.rs
//! Alert/disclosure resolution.
//!
//! Mirrors TS from `pipeline/evaluation/alert.ts`.

use serde::{Deserialize, Serialize};

use crate::types::meaning::Disclosure;

/// How a bid should be annotated for opponents.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub enum AnnotationType {
    Alert,
    Announce,
    Educational,
}

/// Alert resolution result.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct BidAlert {
    pub alertable: bool,
    pub annotation_type: Option<AnnotationType>,
}


/// Resolve a BidAlert from a disclosure level.
pub fn resolve_alert(disclosure: Disclosure) -> BidAlert {
    match disclosure {
        Disclosure::Alert => BidAlert {
            alertable: true,
            annotation_type: Some(AnnotationType::Alert),
        },
        Disclosure::Announcement => BidAlert {
            alertable: true,
            annotation_type: Some(AnnotationType::Announce),
        },
        Disclosure::Natural | Disclosure::Standard => BidAlert {
            alertable: false,
            annotation_type: Some(AnnotationType::Educational),
        },
    }
}

